File: problems/452_minimum-number-of-arrows-to-burst-balloons/main.cpp
#include <iostream>
#include <sstream>
#include <vector>

using namespace std;

class Solution
{
public:
    int findMinArrowShots(vector<vector<int>> &points)
    {
        sort(points.begin(), points.end(), [](const vector<int> &a, const vector<int> &b)
             { return a[1] < b[1]; });

        int arrows = 1;
        int prevEnd = points[0][1];

        // Count the number of non-overlapping intervals
        for (int i = 1; i < points.size(); ++i)
        {
            if (points[i][0] > prevEnd)
            {
                arrows++;
                prevEnd = points[i][1];
            }
        }

        return arrows;
    }
};

string vec2d_to_string(const vector<vector<int>> &data)
{
    stringstream ss;
    ss << "[";
    for (size_t i = 0; i < data.size(); ++i)
    {
        if (i > 0)
        {
            ss << ", ";
        }
        ss << "[" << data[i][0] << ", " << data[i][1] << "]";
    }
    ss << "]";
    return ss.str();
}

int main()
{
    Solution sol;

    vector<vector<vector<int>>> samples = {
        {{10, 16}, {2, 8}, {1, 6}, {7, 12}},
        {{1, 2}, {3, 4}, {5, 6}, {7, 8}},
        {{1, 2}, {2, 3}, {3, 4}, {4, 5}}};
    vector<int> expected = {2, 4, 2};

    for (size_t i = 0; i < samples.size(); ++i)
    {
        vector<vector<int>> points = samples[i];
        int result = sol.findMinArrowShots(points);

        cout << "Sample " << (i + 1) << ": points = " << vec2d_to_string(samples[i]) << "\n";
        cout << "Placeholder result: " << result << "\n";
        cout << "Expected output: " << expected[i] << "\n\n";
    }

    return 0;
}

File: problems/452_minimum-number-of-arrows-to-burst-balloons/main.rs
struct Solution;

impl Solution {
    pub fn find_min_arrow_shots(mut points: Vec<Vec<i32>>) -> i32 {
        if points.is_empty() {
            return 0;
        }
        points.sort_by(|a, b| a[1].cmp(&b[1]));
        let mut arrows = 1;
        let mut prev_end = points[0][1];
        for i in 1..points.len() {
            if points[i][0] > prev_end {
                arrows += 1;
                prev_end = points[i][1];
            }
        }
        arrows
    }
}

fn main() {
    let samples = vec![
        (vec![vec![10, 16], vec![2, 8], vec![1, 6], vec![7, 12]], 2),
        (vec![vec![1, 2], vec![3, 4], vec![5, 6], vec![7, 8]], 4),
        (vec![vec![1, 2], vec![2, 3], vec![3, 4], vec![4, 5]], 2),
    ];

    for (idx, (points, expected)) in samples.into_iter().enumerate() {
        let result = Solution::find_min_arrow_shots(points.clone());

        println!("Sample {}: points = {:?}", idx + 1, points);
        println!("Placeholder result: {}", result);
        println!("Expected output: {}\\n", expected);
    }
}
